### PR TITLE
chore: cache half secret http calls

### LINKED ISSUE #126: cache half secret HTTP calls
temporarily cache half secret HTTP calls

### PR DESCRIPTION
**Pull request by Baha-sk**
closes #126

Signed-off-by: Baha Shaaban <baha.shaaban@securekey.com>

### CODE REVIEW

File: pkg/secretlock/secretsplitlock/secret_split_lock.go [hunk 1]
```diff
@@ -59,15 +68,37 @@ func New(secret []byte, params *HubAuthParams, options ...Option) (secretlock.Se
 		options[i](opts)
 	}
 
+	doOnce.Do(func() {
+		opts.Logger.Infof("Initializing cache...")
+		cache = gcache.New(0).
+			LoaderFunc(func(key interface{}) (interface{}, error) {
+				uriAndToken := strings.Split(fmt.Sprintf("%s", key), "&")
+
+				return fetchSecretShare(uriAndToken[0], uriAndToken[1], opts.HTTPClient, opts.Logger)
```
Comment (Baha-sk) [inline pkg/secretlock/secretsplitlock/secret_split_lock.go:77]: `opts.HTTPClient` and `opts.Logger` are registered for `doOnce.Do` and therefore do not change for subsequent calls. This is why unit-tests with custom http client are commented out.

File: pkg/secretlock/secretsplitlock/secret_split_lock.go [hunk 2]
```diff
@@ -59,15 +68,37 @@ func New(secret []byte, params *HubAuthParams, options ...Option) (secretlock.Se
...
+
 	if secret == nil {
 		return nil, errors.New("empty secret share")
 	}
 
-	otherSecret, err := fetchSecretShare(params.URL, params.APIToken, params.Subject, opts.HTTPClient, opts.Logger)
+	uri := fmt.Sprintf("%s%s?sub=%s", params.URL, hubAuthSecretPath, url.QueryEscape(params.Subject))
+	token := fmt.Sprintf("Bearer %s", base64.StdEncoding.EncodeToString([]byte(params.APIToken)))
+
+	cachedSecret, err := cache.Get(uri + "&" + token)
```
Comment (llorllale) [inline pkg/secretlock/secretsplitlock/secret_split_lock.go:92]: Using `&` as a delimiter is a bad idea when one of the components is a URI.

It's safer if you first `url.PathEscape(hubAuthSecretPath)` and then use a [non-reserved](https://tools.ietf.org/html/rfc3986#section-2.2) character as delimiter (eg. `|`).

**Review by llorllale — APPROVED** [review]

File: pkg/secretlock/secretsplitlock/secret_split_lock.go (hunk 2 above)
Comment (bstasyszyn) [reply pkg/secretlock/secretsplitlock/secret_split_lock.go:92]: Using a struct as a key also works:

type cacheKey struct {
    uri   string
    token string
}

**Comment by aholovko** [comment]
Closed in favor of https://github.com/trustbloc/hub-kms/pull/132

**Pull request closed without merging**